Joao gabriel abreu

joaogsma opened the pull request.

albertaouad commented:
Hi @joaogsma,

I apologize for the delay in reviewing your challenge.
Can you please provide us with the following info:

- Steps to Build/Test/Run
- Diagram about your architecture and maybe why the actor model
- An explanation of how does this solution scale

joaogsma commented:
Hi @albertaouad,

The steps to build/test/run I added to the README, and the class diagram I added as a new file.

As to why I chose the actor model, there are a couple of reasons. First, I simply wanted to try out the Akka framework. Also, I think the actor model is a natural choice for this problem, because each cat can be modeled as a different actor. It is also considered one of the best ways to model concurrency problems, and it is the standard way in Scala to solve concurrency, besides `Futures`. In fact, since Scala 2.10, the language deprecated its actor model implementation and [encourages users to use Akka's](https://docs.scala-lang.org/overviews/core/actors-migration-guide.html).

As to why the solution scales, Akka's actors are lightweight, and a single machine can support millions of then. Also, the framework supports working in cluster mode, in which the actors are distributed across the machines. I haven't read that part of the documentation yet, but it seems it would be relatively simple to migrate this solution to work in a cluster.
